src/System.Private.CoreLib/shared/System/Runtime/Intrinsics/X86/Bmi2.PlatformNotSupported.cs

```diff
@@ -61,8 +70,16 @@ public abstract class X64
         /// <summary>
         /// unsigned int _mulx_u32 (unsigned int a, unsigned int b, unsigned int* hi)
         ///   MULX r32a, r32b, reg/m32
+        /// The above native signature does not directly correspond to the managed signature. 
+        /// </summary>
+        public static uint MultiplyNoFlags(uint left, uint right) { throw new PlatformNotSupportedException(); }
+
+        /// <summary>
+        /// unsigned int _mulx_u32 (unsigned int a, unsigned int b, unsigned int* hi)
+        ///   MULX r32a, r32b, reg/m32
+        /// The above native signature does not directly correspond to the managed signature. 
         /// </summary>
-        public static unsafe uint MultiplyNoFlags(uint left, uint right, uint* high) { throw new PlatformNotSupportedException(); }
+        public static unsafe uint MultiplyNoFlags(uint left, uint right, uint* low) { throw new PlatformNotSupportedException(); }
```

Comment:
Erm, implementing this intrinsic will be fun.

Reply:
@mikedn Concerns?

Reply:
I think it's the same concern raised in the other thread, in that (like with Div/Rem) we don't necessarily have an easy way to return multiple results from a single instruction.

I think the "inefficient" form might be easier, but it will require an explicit load to memory as part of the codegen (where-as the efficient one would let us keep it in register/etc).

Reply:
> think the "inefficient" codegen might be easier, but it will require an explicit store to 

The perf of "explicit store" would be fine, `uint`/`ulong` can be handled by store forward on Intel CPUs. And  we have provided "no store" version, which avoids additional stores when users do not need "low".

Reply:
This intrinsic both stores to memory and returns a value. That's a bit unusual, I think the only similar opers are the atomic ones. We'll see how it goes.

Reply:
Ah, thanks for reminding.

Reply:
> The perf of "explicit store" would be fine, uint/ulong can be handled by store forward on Intel CPUs.

It depends what people need this for. For those that are only after the high part this should be fine. Those that want the low part will probably still pay a few cycles even with store forwarding.

> And we have provided "no store" version, which avoids additional stores when users do not need "low".

That one is useless.

Reply:
It's not useless - when doing division by multiplying with the reciprocal, then "low" is ignored.

Reply:
Ah, the other one still returns the high part. I thought it is returning only the low part.

Well, it's still kind of useless, or better said inconvenient. This particular functionality exists on any x86/x64 CPU, but it is exposed only as part of BMI2.

Reply:
Yes, it's a bit unfortunate that we don't have `X86.Base` like ARM does, but hopefully we'll get there. I'm really hoping to use 128-bit DIV & MUL one day!

Reply:
> Well, it's still kind of useless, or better said inconvenient. This particular functionality exists on any x86/x64 CPU, but it is exposed only as part of BMI2

The BMI2 specific instruction, which doesn't set any CPU flags will be part of BMI2 exclusively.

There is a separate discussion about exposing additional `Base` intrinsics (for things like `mul`, `div`, etc) that needs a proposal, review, etc
* We probably want "general-purpose" versions exposed in `System.Math` (and for existing methods to be special-cases, where applicable). However, there are still special semantics to the x86 instructions that may not be generally applicable and for which exposing intrinsics would still be desirable

Reply:
Also, to be clear, the `Base` intrinsics would not apply to just any instruction; but for ones which expose special semantics that are desirable for high-perf scenarios (much as is done for the C/C++ intrinsics that fit this category)

Reply:
> The BMI2 specific instruction, which doesn't set any CPU flags will be part of BMI2 exclusively.

Apparently you are having trouble understanding that nobody needs or asked for the "no flags" functionality while everyone asked for the broadly available 128 bit functionality. And instead of that they've got the somewhat less available BMI2 form.

Reply:
> while everyone asked for the broadly available 128 bit functionality

No, I completely understood that users have wanted the more broadly available functionality, which is why I mentioned the existing discussions that have been had for the other requested functionality.

> And instead of that they've got the somewhat less available BMI2 form.

Yes, they will get this initially, because it is part of the already reviewed/approved API set.

Adding APIs for things like `__addcarry`, `_bittest`, `_mul128`, `_mulh`, etc... were not part of the initially reviewed/approved API set and have to go through the process independently (and are tracked by issues such as https://github.com/dotnet/corefx/issues/32075, and others which are scattered about).